### DISCUSSION TITLE
How to extract the connection string and pass it as program argument

### DISCUSSION THREAD

**Discussion by Kralizek** [microsoft/aspire#4172] [Q&A] [answered]
I'm stitching together different snippets from multiple sources (discussions, twitter, reddit and so on) and I feel I'm getting closer and closer to get my application being fully built on Aspire despite many components not being super friendly to its model.

Now I'm at the stage that I want to run a console application we use to initialize and seed our database. This operation is idempotent, so I don't mind adding it to the AppHost. Unfortunately, it needs the connection string to the database to be passed as an argument and I can't find a way to get it.

```csharp
using AppHost.Extensions;
using Projects;

var builder = DistributedApplication.CreateBuilder(args);

var adminUiClientId = builder.AddParameter("AdminUiClient");
var adminUiClientSecret = builder.AddParameter("AdminUiClientSecret");
var adminUiLicenseKey = builder.AddParameter("AdminUiLicenseKey");

var sqlPasswordResource = builder.AddParameter("SqlServerPassword");

var mssql = builder.AddSqlServer("db", sqlPasswordResource)
    .PublishAsConnectionString()
    .WithHealthCheck()
    .WithDataVolume();

var db = mssql.AddDatabase("Database", "identity-manager")
    .AddConnectionStringData(new Dictionary<string, object>
    {
        ["MultipleActiveResultSets"] = true,
        ["Encrypt"] = false
    }); // source: https://github.com/dotnet/aspire/discussions/3605

var legacyDb = mssql.AddDatabase("LegacyDatabase", "legacy")
    .AddConnectionStringData(new Dictionary<string, object>
    {
        ["MultipleActiveResultSets"] = true,
        ["Encrypt"] = false
    });

var legacy = builder.AddProject<LegacyAPI>("legacy", "https")
    .WaitFor(mssql) // source: https://github.com/davidfowl/WaitForDependenciesAspire
    .WithReference(db);

var auth = builder.AddProject<AuthService>("auth", "https")
    .WaitFor(mssql)
    .WithReference(db)
    .WithReference(legacyDb);

const string protocol = "https";

var dbOpsRunner = builder.AddProject("dbOpsRunner", "../DatabaseOpsRunner/DatabaseOpsRunner.csproj")
    .WaitFor(mssql)
    .WithArgs(["apply", "all", $"{db}"]) // insert connection string here
    .WithEnvironment("ADMINUI__ADMINUICLIENTID", adminUiClientId)
    .WithEnvironment("ADMINUI__UIURL", "https://localhost:7089") // add endpoint to auth
    .WithEnvironment("ADMINUI__AUTHORITYURL", auth.GetEndpoint(protocol))
    .WithEnvironment("ADMINUI__LICENSEKEY", adminUiLicenseKey);

EndpointReference adminEndpoint = default!;

var admin = builder.AddProject<Admin>("admin", "https")
    .WaitForCompletion(dbOpsRunner)
    .WaitFor(mssql)
    .WithEnvironment("AdminUi__AdminUiClientId", adminUiClientId)
    .WithEnvironment("AdminUi__AdminUiClientSecret", adminUiClientSecret)
    .WithEnvironment("AdminUi__LicenseKey", adminUiLicenseKey)
    .WithEnvironment("AdminUi__AuthorityUrl", auth.GetEndpoint(protocol))
    .WithEnvironment("AdminUi__UiUrl", () => adminEndpoint.Url)
    .WithReference(db);

adminEndpoint = admin.GetEndpoint(protocol);

// source: https://github.com/davidfowl/AspireYarp
// builder.AddYarp("login")
//     .Route("legacy", legacy, path: "/accounts", preservePath: true)
//     .Route("auth", auth, path: "/");
// .WithHttpEndpoint(port: 8001)
// .WithReference(auth)
// .LoadFromConfiguration("ReverseProxy");

builder.Build().Run();
```

I tried the following:

```csharp
.WithArgs(["apply", "all", $"{db}"])

// Result: the whole object gets printed out.
```

```csharp
.WithArgs(["apply", "all", $"{await db.Resource.GetConnectionStringAsync()}"])

// Result:

Unhandled exception. System.InvalidOperationException: The endpoint `tcp` is not allocated for the resource `db`.
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_AllocatedEndpoint() in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 107
   at Aspire.Hosting.ApplicationModel.EndpointReference.get_Port() in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 79
   at Aspire.Hosting.ApplicationModel.EndpointReferenceExpression.GetValueAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/ApplicationModel/EndpointReference.cs:line 179
   at Aspire.Hosting.ApplicationModel.ReferenceExpression.GetValueAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/ApplicationModel/ReferenceExpression.cs:line 66
   at Aspire.Hosting.ApplicationModel.ReferenceExpression.GetValueAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/ApplicationModel/ReferenceExpression.cs:line 66
   at Aspire.Hosting.ApplicationModel.ReferenceExpression.GetValueAsync(CancellationToken cancellationToken) in /_/src/Aspire.Hosting/ApplicationModel/ReferenceExpression.cs:line 66
   at Program.<Main>$(String[] args) in C:\Development\App\tools\AppHost\Program.cs:line 42
   at Program.<Main>(String[] args)

Process finished with exit code -532,462,766.
```

Another issue I'm having is passing the endpoint to a service to the console app. The tricky part is that I want the service to wait for the console app to terminate before it starts. For now I have hardcoded the port number.

See

```
.WithEnvironment("ADMINUI__UIURL", "https://localhost:7089") // add endpoint to auth
```

**Comment by davidfowl** [comment, marked as answer]
It's not intuitive how you'd make this work:

```C#
.WithArgs(context =>
{
  context.Args.Add("apply");
  context.Args.Add("all");
  context.Args.Add(new ConnectionStringReference(db.Resource, optional: false)));
})
```

We need to add the string interpolation overload to adding args.

> Another issue I'm having is passing the endpoint to a service to the console app. The tricky part is that I want the service to wait for the console app to terminate before it starts. For now I have hardcoded the port number.

I'm not fully understanding this. You want the console app to terminate before which service starts?

**Reply by Kralizek** [reply]
> I'm not fully understanding this. You want the console app to terminate before which service starts?

Sorry for the convoluted sentence.
Yes. But the console app needs to know the endpoint of the admin service to finish the initialization.

**Reply by MarcelChirtes** [reply]
Dumb ask: how can we get the ConnectionString with 127.0.0.1? 
For now I'm going with the classical, it works, we'll see later: `ConnectionString.Replace("localhost", "127.0.0.1");`.

If is there a way to set up the ip based networking instead of hostname from the DistributedApplication builder, I would like that a lot.

Context: I'm new to aspire, and I'm using it for the wrong thing, for now in xUnit to as replacement of: `Process.Start(ProcessStartInfo("docker or podman composer"))` in background. I already love Aspire has a start and stop accessible in tests, managing the resources it's a dream, planing to actually extend tests by replacing mocking http calls with actual resources to csproj or dockerfiles (non dotnet).

**Reply by Kralizek** [reply]
Why don't you use test containers instead? 🤔

**Reply by MarcelChirtes** [reply]
I could, one feature that I like is the fact it has a strategy to wait until Sql Server response, maybe Aspire has one, could not find till today.
Still, what I would hope to achieve with Aspire, at some point to have the Aspire metrics and logs part of the assertions.

**Reply by Kralizek** [reply]
You can wait for a database to be available using Aspirant packages. The problem is that Aspire is not a good fit (imho) for this scenario.

**Reply by MarcelChirtes** [reply]
Understood, appreciate the advice, probably I'll take it, for now I solved all my problems with dirty changes and checks.

**Reply by MarcelChirtes** [reply]
Thanks again!!

**Reply by kwaazaar** [reply]
> ```cs
> .WithArgs(context =>
> {
>   context.Args.Add("apply");
>   context.Args.Add("all");
>   context.Args.Add(new ConnectionStringReference(db.Resource, optional: false)));
> })
> ```

So how do I pass the connectionstring as an environment variable, but with a different name?

(I'm almost ashamed to have to ask something this st00pid, but I really cannot find any docs anywhere)

**Reply by MarcelChirtes** [reply]
> > ```cs
> > .WithArgs(context =>
> > {
> >   context.Args.Add("apply");
> >   context.Args.Add("all");
> >   context.Args.Add(new ConnectionStringReference(db.Resource, optional: false)));
> > })
> > ```
> 
> So how do I pass the connectionstring as an environment variable, but with a different name?
> 
> (I'm almost ashamed to have to ask something this st00pid, but I really cannot find any docs anywhere)

maybe:

```
builder.AddProject<Projects.MyWebApp>("webapp")
    .WithReference(db)
    .WithEnvironment("ConnectionStrings__MyConnection", 
        new ConnectionStringReference(db.Resource, optional: false))
    .WithEnvironment("SOMETHINGELSE", 
        new ConnectionStringReference(db.Resource, optional: false));
```

This overrides whatever is in appsettings at
```
{
  "ConnectionStrings": {
    "MyConnection": "Server=(localdb)\\mssqllocaldb;Database=Test;"
  },
}
```
Accessible in `MyWebApp` with
```
 builder.Configuration.GetConnectionString("MyConnection")`
```
Or with 
```
string connectionStringsMyConnection = Environment.GetEnvironmentVariable("ConnectionStrings__MyConnection");
string somethingElse = Environment.GetEnvironmentVariable("SOMETHINGELSE");
```

**Reply by kwaazaar** [reply]
I used a resource in an interpolated string. But even though it holds just a string, it's ToString does not return its value. So I had to use `$"{customerCode.Resource.Value}__Bus__ConnectionString"` instead of `$"{customerCode}__Bus__ConnectionString"`. Took me hours to figure this out.

So basically it's pretty straightforward and you don't need `ConnectionStringReference` (anymore):

```
var builder = DistributedApplication.CreateBuilder(args);

var customerCode = builder.AddParameter("customerCode"); // this is a Resource, not a string!
var saPassword = builder.AddParameter("saPassword", secret: true);

var sqlServer = builder
    .AddSqlServer("sql", password: saPassword, port: 14329)
    .WithDataVolume(); // Keep data between restarts

var dbMaster = sqlServer.AddDatabase("master");
var dbCustomer = sqlServer.AddDatabase(customerCode.Resource.Value);

builder
    .AddProject<Projects.MessageProcessor>("MessageProcessor")
    .WithEnvironment($"{customerCode.Resource.Value}__CustomerSettings__DbConnectionString", dbCustomer);

builder.Build().Run();
```

**Reply by davidfowl** [reply]
~This should be fixed in 9.2~, why is the name of the environment variable the value here?

**Reply by kwaazaar** [reply]
`.WithEnvironment` takes name, then value. Are you confusing it with `.WithReference`, which takes value then (connection)name?
Otherwise I'm not sure what you mean.

**Reply by davidfowl** [reply]
WithEnvironment($"{customerCode.Resource.Value}__CustomerSettings__DbConnectionString", dbCustomer);


I see you wanted the value of the parameter to be part of the name of the environment variable.

PS: This won’t be publishable as the parameter has no value at the time this code is run.

**Reply by davidfowl** [reply]
I wrote this doc https://gist.github.com/davidfowl/b408af870d4b5b54a28bf18bffa127e1 (it's BIG), please let me know if it helps with your understanding and feel free to leave any feedback.
